### ISSUE TITLE
Combined loss

### ISSUE THREAD

**Issue by John1231983** [voxelmorph/voxelmorph#98]
As far as I know, the ncc loss returns a negative value. How we combined it with another loss, such as segmentation loss (cross-entropy, dice...)--that is positive value?

I suggest we can use a upper bound value and plus it with ncc to convert it to the positive range, such as
```
L_total = (5+ncc_loss) + dice_loss
```
where `5` is the maximum loss value that NCC can achieve. But the problem is how we can define the upper bound value, it really depends on the dataset.

**Comment by adalca** [comment]
a couple points:

 - how to combine losses where one happens to be positive and one happens to be negative (of course, remember, you want to minimize it)? I don't see a problem with `L_total = ncc_loss + positive_loss`. In fact, `5` in your example would have no effect on training, since the derivative with respect to 5 is 0. 

 - Dice as a measure itself is [0, 1], and you want to maximize this. so a Dice Loss can be `dice_loss = -Dice_metric`, which is actually negative. Now, some code bases might implement it as `dice_loss = 1 - Dice_metric` just to keep it positive, but again, it won't have an effect on training.

**Comment by John1231983** [comment]
@adalca : Thanks. How about the option `L_total = ncc_loss + positive_loss` where ncc loss is negative and positive_loss is cross-entropy. When ncc_loss goes down such as -0.2, then cross entropy loss is 0.1 then L_total is -0.1. Do you think the direction of gradient will change to the direction of ncc loss (reg loss)

**Comment by adalca** [comment]
no, the gradient direction is not affected by the sign of the loss
That is, if you have `5 + neg_loss` vs `neg_loss`, the gradient will be the same.

**Comment by John1231983** [comment]
But normally, adding the new loss (i.e. ncc loss), the total loss must be increase in comparison without adding the new loss. But in the case of ncc loss, the total loss is reduced (because of ncc loss is negative). If we use MSE loss, then it is fine. However, I found that ncc loss works better MSE in your network

**Comment by adalca** [comment]
> But normally, adding the new loss (i.e. ncc loss), the total loss must be increase in comparison without adding the new loss. 

I don't understand this, why would this be?

**Comment by John1231983** [comment]
@adalca : For example, mse loss is a new loss that added to the cross-entropy loss, then the total loss will be bigger because of mse always bigger than zero.

For ncc loss, it often smaller than zero, so the total loss will be decreased. In registration, we can use both mse or ncc, so if we add mse or ncc loss to the total loss, then the total loss is very different in comparison with mse and ncc (mse increase the total loss, while ncc decrease the total loss)

**Comment by adalca** [comment]
yes, sorry, I understand that -- you said 

> the total loss **must** be increase

I'm asking why you think the total loss **has to** increase? Yes, adding a negative loss will decrease the total loss -- why would this be an issue. The training will still try to decrease the total loss, by decreasing each term.

**Comment by John1231983** [comment]
Let consider three cases:

- case 1: L_total = cross_entropy loss
- case 2: L_total = cross_entropy loss + mse loss 
- case 3: L_total = cross_entropy loss + ncc loss

The L_total (1) < L_total (2), while L_total (1) > L_total(3). 

During training, I agree that all loss will be decreased. However, the L_total(3) smaller than L_total(1) (original loss), that means the update process will be slower because
```
w_t+1 = w_t - alpha*gradL_total(.)/grad(w)
```

**Comment by adalca** [comment]
perhaps I'm not sure what you're asking -- are you asking whether adding another loss to cross_entropy change the gradient? yes, it will certainly change the gradient, and hence updates will be different. 

What I was trying to say (re: your original question) is that *the fact* that ncc is negative is not a concern -- you can simply add it as another loss.

**Comment by happybcloud** [comment]
Would you please share the code that combined loss with segmentation (cross-entropy, dice...)-

**Comment by adalca** [comment]
the Dice loss is available in the losses.py file. 

Or do you mean the whole semi-supervision part of the network? if so, This is available in the `redesign` branch under `VxmDenseSegSemiSupervised` -- this branch will be merged into the master soon, but can still be used if you just switch the branch in your current repository clone.
